packages/puppeteer-core/src/common/bidi/Serializer.ts

```diff
@@ -0,0 +1,226 @@
+import {Bidi} from '../../../third_party/chromium-bidi/index.js';
+import {debugError, isPlainObject} from '../util.js';
+
+class UnserializableError extends Error {}
+
+/**
+ * @internal
+ */
+export class BidiSerializer {
+  static serializePrimitive(
+    arg: unknown
+  ): Bidi.CommonDataTypes.PrimitiveProtocolValue {
+    const type = typeof arg as Exclude<
+      Bidi.CommonDataTypes.PrimitiveProtocolValue['type'],
+      'null'
+    >;
+
+    switch (type) {
+      case 'undefined':
+        return {
+          type: 'undefined',
+        };
+      case 'number': {
+        let value!: Bidi.CommonDataTypes.SpecialNumber | number;
+        if (Object.is(arg, -0)) {
+          value = '-0';
+        } else if (Object.is(arg, Infinity)) {
+          value = 'Infinity';
+        } else if (Object.is(arg, -Infinity)) {
+          value = '-Infinity';
+        } else if (Object.is(arg, NaN)) {
+          value = 'NaN';
+        } else {
+          value = arg as number;
+        }
+        return {
+          type,
+          value,
+        };
+      }
+      case 'bigint':
+        return {
+          type,
+          value: (arg as bigint).toString(),
+        };
+      case 'string': {
+        return {
+          type,
+          value: arg as string,
+        };
+      }
+      case 'boolean':
+        return {
+          type,
+          value: arg as boolean,
+        };
+
+      default:
+        throw new Error(
+          `Primitive data of type ${typeof arg} and value ${arg} not supported for BiDi serialization`
+        );
+    }
+  }
+
+  static serializeRemoveValue(
+    arg: unknown
+  ): Bidi.CommonDataTypes.LocalOrRemoteValue {
+    switch (typeof arg) {
+      case 'symbol':
+      case 'function':
+        throw new UnserializableError('Unable to serializable input.');
+      case 'object':
+        if (arg === null) {
+          return {
+            type: 'null',
+          };
+        } else if (Array.isArray(arg)) {
+          const parsedArray = arg.map(subArg => {
+            return BidiSerializer.serializeRemoveValue(subArg);
+          });
+
+          return {
+            type: 'array',
+            value: parsedArray,
+          };
+        } else if (isPlainObject(arg)) {
+          const parsedObject: Bidi.CommonDataTypes.MappingLocalValue = [];
+          for (const key in arg) {
+            parsedObject.push([
+              BidiSerializer.serializePrimitive(key),
+              BidiSerializer.serializeRemoveValue(arg[key]),
+            ]);
+          }
+
+          return {
+            type: 'object',
+            value: parsedObject,
+          };
+        }
+
+        throw new UnserializableError(
+          'Custom object sterilization not possible. Use plain objects instead.'
+        );
+
+      case 'undefined':
```

Comment:
Maybe it would be better to move all Primitive here as well, and have support functions for Object and number, WDYT?